pulsar-functions/transforms/src/main/java/org/apache/pulsar/functions/transforms/TransformContext.java

```diff
@@ -0,0 +1,131 @@
+            KeyValue kv = (KeyValue) value;
+            this.keySchema = kvSchema.getKeySchema();
+            this.keyObject = this.keySchema.getSchemaInfo().getType() == SchemaType.AVRO
+                    ? ((org.apache.pulsar.client.api.schema.GenericRecord) kv.getKey()).getNativeObject()
+                    : kv.getKey();
+            this.valueSchema = kvSchema.getValueSchema();
+            this.valueObject = this.valueSchema.getSchemaInfo().getType() == SchemaType.AVRO
+                    ? ((org.apache.pulsar.client.api.schema.GenericRecord) kv.getValue()).getNativeObject()
+                    : kv.getValue();
+            this.keyValueEncodingType = kvSchema.getKeyValueEncodingType();
+        } else {
+            this.valueSchema = schema;
+            this.valueObject = value;
+            this.key = currentRecord.getKey().orElse(null);
+        }
+    }
+
+    public void send() throws IOException {
+        if (keyModified && keySchema != null && keySchema.getSchemaInfo().getType() == SchemaType.AVRO) {
+            GenericRecord genericRecord = (GenericRecord) keyObject;
+            keySchema = Schema.NATIVE_AVRO(genericRecord.getSchema());
+            keyObject = serializeGenericRecord(genericRecord);
+        }
+        if (valueModified && valueSchema != null && valueSchema.getSchemaInfo().getType() == SchemaType.AVRO) {
+            GenericRecord genericRecord = (GenericRecord) valueObject;
+            valueSchema = Schema.NATIVE_AVRO(genericRecord.getSchema());
+            valueObject = serializeGenericRecord(genericRecord);
+        }
+
+        Schema outputSchema;
+        Object outputObject;
+        GenericObject recordValue = (GenericObject) context.getCurrentRecord().getValue();
+        if (keySchema != null) {
+            outputSchema = Schema.KeyValue(keySchema, valueSchema, keyValueEncodingType);
+            Object outputKeyObject = !keyModified && keySchema.getSchemaInfo().getType().isStruct()
+                    ? ((KeyValue) recordValue.getNativeObject()).getKey()
+                    : keyObject;
+            Object outputValueObject = !valueModified && valueSchema.getSchemaInfo().getType().isStruct()
+                    ? ((KeyValue) recordValue.getNativeObject()).getValue()
+                    : valueObject;
+            outputObject = new KeyValue(outputKeyObject, outputValueObject);
+        } else {
+            outputSchema = valueSchema;
+            outputObject = !valueModified && valueSchema.getSchemaInfo().getType().isStruct()
+                    ? recordValue
+                    : valueObject;
+        }
+
+        if (log.isDebugEnabled()) {
+            log.debug("output {} schema {}", outputObject, outputSchema);
+        }
+        TypedMessageBuilder<?> message = context.newOutputMessage(outputTopic, outputSchema)
+                .properties(properties)
+                .value(outputObject);
+        if (key != null) {
```

Comment:
if we are using KeyValue Schema we should not set the Key this way.

Reply:
Indeed

Reply:
done